### PR TITLE
Add source_chunk_ids overlap filter to list_entities across storage backends

### LINKED ISSUE #1448: list_entities: source_chunk_ids overlap filter (push #857 recall projection into the store)
## Context

The #857 simple-path projection in VectorCypher recall fetches up to 1,000 entities for the **whole namespace** (`list_entities(namespace_id, limit=1000)`) and filters them in Python by overlap with the recalled chunk ids. This is wasteful (full entities incl. embeddings over the wire) and silently wrong on namespaces with >1,000 entities (the cap drops arbitrary entities relative to the recalled set).

Decision: **no new protocol methods** — extend the existing `list_entities` with a `source_chunk_ids` filter kwarg, mirroring the existing `entity_type` kwarg pattern.

## Semantics

New keyword-only parameter on `list_entities`: `source_chunk_ids: list[UUID] | None = None`.

- `None` (default): no filter — behavior byte-identical to today.
- Non-empty list: return only entities whose `source_chunk_ids` contains **at least one** of the given ids (any-overlap, not subset).
- Empty list `[]`: return `[]` (matches nothing) via an early return at the top of each backend.
- Composes with AND against existing conditions (`entity_type`, live/`valid_until` filters) and applies **before** `limit`/`offset`.

## Scope

Extend `list_entities` across the protocol, coordinator, and all backends (Neo4j, Memgraph, pgvector, sqlite_lance, SurrealDB, Neptune, AGE) plus the #857 retriever call site. Backend behavior tests for sqlite_lance/pgvector/Neo4j/SurrealDB + a retriever unit test.

The relationships half (`list_relationships` / `between_entity_ids`) is a **separate upcoming issue — not implemented here**.

### PR DESCRIPTION
**Pull request by filippetrovic**
## Summary

The VectorCypher simple-path (graph-less) recall projection fetched up to 1,000 entities for the **whole namespace** and filtered them in Python by overlap with the recalled chunk ids. That is wasteful (full entities incl. embeddings over the wire) and silently wrong on namespaces with more than 1,000 entities (the cap drops arbitrary entities relative to the recalled set).

This pushes the entity projection into the store. `list_entities` gains a keyword-only filter, mirroring the existing `entity_type` kwarg:

```python
list_entities(namespace_id, *, source_chunk_ids: list[UUID] | None = None, ...)
```

- `None` (default): no filter — behavior **byte-identical** to before.
- Non-empty list: return only entities whose `source_chunk_ids` overlaps **at least one** of the given ids (any-overlap, not subset).
- Empty list `[]`: return `[]` (matches nothing).

The filter composes with `AND` against existing conditions (`entity_type`, live/`valid_until`) and applies **before** `limit`/`offset`.

### Changes
- **Protocol** (`storage/backends/base.py`) and **Coordinator** (`storage/coordinator.py`, both graph + vector-fallback branches) carry the new kwarg.
- **Pushdown backends:** Neo4j / Memgraph (`any(cid IN e.source_chunk_ids ...)`), pgvector (`&&` array-overlap via `.overlap()`), sqlite_lance (`EXISTS json_each`), SurrealDB (`CONTAINSANY`).
- **Neptune / AGE:** predicate pushdown is unverified on those openCypher subsets, so they filter in-method (fetch without `SKIP`/`LIMIT`, Python set-overlap, then slice) — same page and ordering as the pushdown backends.
- **Retriever** (`engines/vectorcypher/retriever.py`, simple-path projection): now passes `source_chunk_ids=list(recalled_chunk_ids)` instead of listing the whole namespace; the Python scoring loop is unchanged.

Relationship-side projection (`list_relationships` / `between_entity_ids`) is intentionally left as a follow-up. No new spans or metrics.

## Test plan
- [x] Unit tests pass (`make test`: 478 passed, 257 skipped, 85% coverage)
- [x] `make format` / `make lint` clean
- [x] Backend behavior tests added for sqlite_lance, pgvector, Neo4j, SurrealDB — each pins the 4-case contract (no filter → both; one chunk → owning entity; unknown id → `[]`; `[]` → `[]`)
- [x] Retriever unit test asserts the recalled chunk ids are pushed down to `list_entities`
- [ ] Docker-backed pgvector/Neo4j integration lanes run in CI

Fixes #1448

## Summary by CodeRabbit

* **New Features**
  * Entity listings can now be filtered by their originating chunk IDs via a new `source_chunk_ids` option, propagated through the main storage layer and supported backends.

* **Bug Fixes**
  * Improved retrieval accuracy by narrowing entity results to the relevant chunk provenance.
  * Passing an empty `source_chunk_ids` list now returns no results immediately.

* **Tests**
  * Added/updated unit and integration coverage to validate filtering behavior across backends and retrieval flows.

### CODE REVIEW

**Comment by filippetrovic** [comment]
## Multi-perspective review — consolidated

Reviewed by a 7-role team (security, performance, correctness, database, test, docs, devil's-advocate) plus a dedicated security audit. **Overall: APPROVE.** The change is correct, backward-compatible (None-default path is byte-identical), well-tested, and a strict improvement over the previous namespace-wide list + Python filter. No blocking correctness or security issues. The items below are follow-ups and notes — several proposed "HIGH" changes are intentionally **out of scope** for this change (documented inline).

### Security audit — PASS
No injection surface. Every backend parameterizes `source_chunk_ids`: `$source_chunk_ids` (Neo4j/Memgraph/SurrealDB), positional `?` with values bound out-of-band (sqlite_lance — the `# noqa: S608` interpolates only `?` placeholders), `.overlap()` bind param (pgvector), and post-fetch Python set-overlap (Neptune/AGE, no query surface). The AGE `page_clause` f-string interpolates only integer `offset`/`limit` (pre-existing pattern). No secrets, auth, crypto, SSRF, deserialization, or new-dependency surface.

### HIGH
- **[Performance] Recommended fast-follow (not in this change's scope):** `pgvector.py` — the `&&` array-overlap (`.overlap()`) benefits from a GIN index on `entities.source_chunk_ids`; without it the filter is a seq scan on large tables. Note this change is still a net win without the index (it stops shipping ~1000 full entities + embeddings over the wire and fixes the >1000-entity correctness bug). A GIN index is a separate migration and deliberately not bundled here — recommend a follow-up: `CREATE INDEX CONCURRENTLY ... USING GIN (source_chunk_ids)`.

### MEDIUM
- **[Performance/Design] Neptune/AGE in-method filter** (`neptune.py`, `age.py`): when `source_chunk_ids` is set these fetch the namespace entity set (no `SKIP`/`LIMIT`), filter in Python, then slice. Correct and deliberate (predicate pushdown is unverified on those openCypher subsets; both carry the `#1448` comment and neither has a CI lane), but it is a full-set scan per recall on large namespaces. Worth a one-line note in the backend docs as a known limitation. (A runtime warning/metric was suggested but is intentionally omitted — no new spans/metrics in this change.)
- **[Test] Composition coverage gap:** no test exercises `source_chunk_ids` combined with `entity_type` (the WHERE clauses compose correctly by inspection). A one-line variant per backend would raise confidence — non-blocking; beyond the pinned 4-case contract.

### LOW
- **[Correctness/Style]** `pgvector.py` — the filter block sits after the `entity_type` clause; harmless, minor placement inconsistency vs other backends.
- **[Docs]** A short comment on the Neptune/AGE fetch-all-then-slice tradeoff (correctness relies on the stable `ORDER BY e.name` + order-preserving filter) would aid future maintainers.
- **[Test]** Multi-overlap edge case (both A & B share a chunk → both returned) is untested; the 4-case contract is minimal but sufficient.
- **[Design/Info]** The now-redundant `overlap > 0` guard in the retriever scoring loop is intentionally left in place (the change's spec calls it "redundant but harmless"); it is not dead-code that needs removal.

### Resolved / non-issues
- **SurrealDB CI lane** (raised as CRITICAL by the test reviewer from diff-only visibility): the PR **is** gated by SurrealDB CI — `conformance (surrealdb)` and `skeleton_surrealdb` lanes run on this PR. Not a gap.
- **Empty-list `return []` repeated per backend:** by design — each backend is independently correct; the early return short-circuits before any query build.
- **Kwarg vs. new `list_entities_by_chunk_ids` method:** deliberate ("no new protocol methods"); mirrors the existing `entity_type` kwarg pattern. Defensible and non-blocking.

_Verdict: APPROVE — merge once CI is green; GIN index tracked as a follow-up._

File: src/khora/storage/backends/sqlite_lance/graph.py [hunk 1]
```diff
@@ -423,6 +424,14 @@ async def list_entities(
...
             conditions.append("entity_type = ?")
             params.append(entity_type)
+        if source_chunk_ids is not None:
+            if not source_chunk_ids:
+                return []
+            placeholders = ", ".join("?" for _ in source_chunk_ids)
+            conditions.append(
+                f"EXISTS (SELECT 1 FROM json_each(entities.source_chunk_ids) WHERE json_each.value IN ({placeholders}))"  # noqa: S608
+            )
+            params.extend(uuid_to_text(c) for c in source_chunk_ids)
```
[the remark this answers is not in the document]
Comment (filippetrovic) [reply src/khora/storage/backends/sqlite_lance/graph.py]: Fixed in 3f134521. The per-id `IN (?, ?, …)` placeholder expansion is replaced with a single bound JSON array:

```sql
EXISTS (SELECT 1 FROM json_each(entities.source_chunk_ids)
        WHERE json_each.value IN (SELECT value FROM json_each(?)))
```

bound with `to_json_text([uuid_to_text(c) for c in source_chunk_ids])`. This removes the SQLITE_MAX_VARIABLE_NUMBER failure mode on wide recalls and keeps the SQL shape constant. Behavior is unchanged — the 4-case sqlite_lance test still passes.

**Review by probablyrobot — APPROVED** [review]
